BleWinrtDll/BleWinrtDll.cpp

```diff
@@ -561,60 +559,201 @@ bool SendData(BLEData* data, bool block) {
+	uint32_t valueCount = 0;
+	uint32_t readIndex = -1;
+	BLECharacteristic characteristics;
+	bool done = false;
+public:
+
+	wchar_t* getDevice() { return this->characteristics.deviceId; }
+	wchar_t* getService() { return this->characteristics.serviceUuid; }
+	wchar_t* getCharacteristic() { return this->characteristics.characteristicUuid; }
+
+	bool operator<(Reading& other) {
+		int val = wcscmp(this->characteristics.deviceId, other.characteristics.deviceId);
+		if (val != 0) return val < 0;
+		val = wcscmp(this->characteristics.serviceUuid, other.characteristics.serviceUuid);
+		if (val != 0) return val < 0;
+		val = wcscmp(this->characteristics.characteristicUuid, other.characteristics.characteristicUuid);
+		return val < 0;
+	}
+
+	void waitDone() {
+		unique_lock valuesLock(valuesMutex);
+		doneSignal.wait(valuesLock, [&]{ return done; });
+	}
+	void finish()
+	{
+		done = true;
+		doneSignal.notify_one();
+	}
+
+	bool operator==(Reading& other) {
+		return *this == other.characteristics;
+	}
+
+	bool operator==(BLECharacteristic& other) {
+		return (
+			(wcscmp(this->characteristics.characteristicUuid, other.characteristicUuid) == 0) &&
+			(wcscmp(this->characteristics.serviceUuid, other.serviceUuid) == 0) &&
+			(wcscmp(this->characteristics.deviceId, other.deviceId) == 0));
+	}
+
+	void reuseRead() {
+		lock_guard readingLock(readingMutex);
+		this->reading = true;
+	}
+
+	bool isReading() {
+		lock_guard readingLock(readingMutex);
+		return this->reading;
+	}
+
+	void stopReading(bool lock = false)
+	{
+		if (lock) {
+			unique_lock readingLock(readingMutex);
+		}
+		else {
+			lock_guard readingLock(readingMutex);
+			this->reading = false;
+		}
+	}
+
+	void setData(uint8_t* data, uint32_t len) {
+		lock_guard lock(valuesMutex);
+		uint64_t l = min(len, sizeof(values));
+		memcpy(values, data, l);
+		++readIndex;
+	}
+
+	void getData(uint8_t* data, uint32_t len)
+	{
+		lock_guard lock(valuesMutex);
+		uint64_t l = min(len, sizeof(values));
+		memcpy(data, values, l);
+	}
+
+};
+
+mutex readingsLock;
+vector<Reading*> readings;
+
+fire_and_forget ReadDataAsync(Reading* reading) {
 	try {
-		auto characteristic = co_await retrieveCharacteristic(id->deviceId, id->serviceUuid, id->characteristicUuid);
-		if (characteristic != nullptr) {
-			//Retrieve data from device not from windows cache
-			auto status = co_await characteristic.ReadValueAsync(BluetoothCacheMode::Uncached);
-			if (status.Status() == GattCommunicationStatus::Success) {
-				auto value = status.Value();
-				auto valueData = value.data();
-				uint32_t length = max(value.Length(), ARRAY_LENGTH(data->buf));
-				for (uint32_t i = 0; i < length; ++i) {
-					data->buf[i] = valueData[i];
+		//auto c = characteristic.CharacteristicProperties();
+		//if (((uint32_t)c & (uint32_t)GattCharacteristicProperties::Read) == 0) {
+		//	saveError(L"%s:%d Characteristic %s does not have read property!", __WFILE__, __LINE__, id->characteristicUuid);
+		//	co_return;
+		//}
```

Comment:
Commented code. As it seems as though we have error handling for getting Characteristic Properties below, I'm assuming this is no longer needed? If not, then remove it.